src/logic/Level.cs

```diff
@@ -493,21 +238,155 @@ public enum VisitedState {
+		public virtual void OnEnter(Level level, Room previousRoom) {
+
+		}
+
+		/// <summary> Called whenever the player exits this room. </summary>
+		public virtual void OnExit(Level level, Room nextRoom) {
+
+		}
+	}
+
+	/// <summary> Room with wind constantly pushing the player around. </summary>
+	public class WindyRoom : Room {
+		public Vector2 WindDirection = new Vector2(0.0f);
+		public float WindSpeed = 0.0f;
+		public float DirectionChangePeriod = 5.0f; //Time between wind direction changes
+		private float _directionChangeTimer = 0.0f;
+		public static readonly string windSoundEffect = "assets/sounds/Wind0.wav";
+
+		public WindyRoom(float x, float y) : base(x, y) { }
+		
+		public override object Clone() {
+			var room = new WindyRoom(Position.X, Position.Y);
+			CloneBase(room);
+			room.WindDirection = WindDirection;
+			room.WindSpeed = WindSpeed;
+			return room;
+		}
+
+		public override void Update(double deltaTime, Level level) {
+			base.Update(deltaTime, level);
+
+			//Push the player around
+			level.Player.Velocity += WindDirection * WindSpeed;
+
+			//Periodically change wind direction
+			_directionChangeTimer += (float)deltaTime;
+			if (_directionChangeTimer >= DirectionChangePeriod) {
+				_directionChangeTimer = 0.0f;
+				WindDirection = new Random().NextVec2();
+			}
+		}
+
+		public override void OnEnter(Level level, Room previousRoom) {
+			base.OnEnter(level, previousRoom);
+			if (!Sounds.IsSoundPlaying(windSoundEffect))
+				Sounds.PlaySound(windSoundEffect); //Start wind sound if it isn't already playing
+		}
+
+		public override void OnExit(Level level, Room nextRoom) {
+			base.OnExit(level, nextRoom);
+			if (nextRoom.GetType() != typeof(WindyRoom))
+				Sounds.StopSound(windSoundEffect); //Stop wind sound if next room isn't windy
+		}
+	}
+
+	
+	/// <summary> Room with slippery floors.  </summary>
+	public class IcyRoom : Room {
+		public float FloorFriction = 0.01f;
+
+		public IcyRoom(float x, float y) : base(x, y) { }
+
+		public override object Clone() {
+			var room = new IcyRoom(Position.X, Position.Y);
+			CloneBase(room);
+			room.FloorFriction = FloorFriction;
+			return room;
+		}
+
+		public override void Update(double deltaTime, Level level) {
+			base.Update(deltaTime, level);
+			base.FloorTexture = "Ice.png";
+		}
+
+		public override void OnEnter(Level level, Room previousRoom) {
+			base.OnEnter(level, previousRoom);
+			//Make the floors slippery and movement slower while in icy rooms
+			level.Player.FloorFriction = FloorFriction;
+			level.Player.MovementSpeed = 0.1f;
+		}
+
+		public override void OnExit(Level level, Room nextRoom) {
+			base.OnExit(level, nextRoom);
+			level.Player.FloorFriction = Player.DefaultFloorFriction;
+			level.Player.MovementSpeed = Player.DefaultMovementSpeed;
+		}
+	}
+
+	/// <summary> Interactable object found in rooms. </summary>
+	public class LevelObject : ICloneable {
+		public Vector2 Position;
+		public readonly string TextureName;
```

Comment:
This should be a reference to some sort of model loading after the model loading pull request is merged.